libtiny_client/src/state.rs

```diff
@@ -232,6 +305,73 @@ impl StateInner {
                 }
             }
 
+            // Reply 477 when user needs to be identified with NickServ to join a channel
+            // ex. Reply { num: 477, params: ["<your_nick>", "<channel name>", "<Server reply message>"] }
+            Reply { num: 477, params } => {
+                // Only try to automatically rejoin if nickserv_ident is configured
+                if let (Some(channel), Some(msg_477)) = (params.get(1), params.get(2)) {
+                    snd_ev
+                        .try_send(Event::Msg(wire::Msg {
+                            pfx: pfx.clone(),
+                            cmd: wire::Cmd::PRIVMSG {
+                                ctcp: None,
+                                is_notice: true,
+                                msg: msg_477.clone(),
+                                target: wire::MsgTarget::Chan(channel.clone()),
+                            },
+                        }))
+                        .unwrap();
+                    // Get channel name from params
+                    if self.nickserv_ident.is_some() {
+                        // Helper for creating an event
+                        let create_message = |msg: String| Event::ChannelJoinError {
+                            chan: channel.clone(),
+                            msg,
+                        };
+                        // Find channel in self.chans
+                        if let Some(idx) = utils::find_idx(&self.chans, |c| c.name == *channel) {
+                            let chan = &mut self.chans[idx];
+                            // Retry joining channel if retries are available
+                            if let Some(retries) = chan.retry_join() {
+                                let retry_msg = format!(
+                                    "Attempting to rejoin {} in 10 seconds... ({}/{})",
+                                    channel,
+                                    MAX_JOIN_RETRIES - retries,
+                                    MAX_JOIN_RETRIES
+                                );
+                                snd_ev.try_send(create_message(retry_msg)).unwrap();
+                                let snd_irc_msg = snd_irc_msg.clone();
+                                // Spawn task and delay rejoin to give NickServ time to identify nick
+                                let (snd_abort, rcv_abort) = tokio::sync::mpsc::channel::<()>(1);
+                                match &mut chan.join_state {
+                                    JoinState::Initial => chan.set_joining(snd_abort),
+                                    JoinState::Joining { stop_task, .. } => *stop_task = snd_abort,
+                                    JoinState::Joined => {
```

Comment:
Wasn't really sure how to handle this unexpected state. Also see below where there is another unexpected state.

Reply:
So basically we got a 477 for a channel that we've already joined. This is weird, yeah. Good idea to log this with ERROR level.